server/index.js

```diff
@@ -42,25 +46,150 @@ var commands = {
+            return;
+        }
+
+        if (!body.app || !body.source || !body.incarnationNumber) {
+            // TODO Typed Error
+            callback(new Error('app, source, incarnationNumber are all required'));
+            return;
+        }
+
+        handleJoin({
+            ringpop: ringpop,
+            app: body.app,
+            source: body.source,
+            incarnationNumber: body.incarnationNumber
+        }, thriftCallback(callback));
+    };
+}
+
+function protocolPingHandler(ringpop) {
+    /* jshint maxparams:5 */
+    return function handleIt(opts, req, head, body, callback) {
+        if (!body) {
+            // TODO Typed error
+            callback(new Error('body is required'));
+            return;
+        }
+
+        // NOTE sourceIncarnationNumber is an optional argument. It was not present
+        // until after the v9.8.12 release.
+        if (!body.changes || !body.checksum || !body.source) {
+            // TODO Typed error
+            callback(new Error('changes, checksum and source are all required'));
+            return;
+        }
+
+        handlePing({
+            ringpop: ringpop,
+            source: body.source,
+            sourceIncarnationNumber: body.sourceIncarnationNumber,
+            changes: body.changes,
+            checksum: body.checksum
+        }, thriftCallback(callback));
+    };
+}
+
+function protocolPingReqHandler(ringpop) {
+    /* jshint maxparams:5 */
+    return function handleIt(opts, req, head, body, callback) {
+        if (!body) {
+            // TODO Typed error
+            callback(new Error('body is required'));
+            return;
+        }
+
+        // NOTE sourceIncarnationNumber is an optional argument. It was not present
+        // until after the v9.8.12 release.
+        if (!body.target || !body.changes || !body.checksum || !body.source) {
+            // TODO Typed error
+            callback(new Error('target, changes, checksum and source are all required'));
+            return;
+        }
+
+        handlePingReq({
+            ringpop: ringpop,
+            source: body.source,
+            sourceIncarnationNumber: body.sourceIncarnationNumber,
+            target: body.target,
+            changes: body.changes,
+            checksum: body.checksum
+        }, thriftCallback(callback));
+    };
+}
+
+function registerThriftHandlers(ringpop, tchannel, tchannelAsThrift) {
+    var thriftHandlers = {
+        'Ringpop::join': protocolJoinHandler(ringpop),
+        'Ringpop::ping': protocolPingHandler(ringpop),
+        'Ringpop::pingReq': protocolPingReqHandler(ringpop)
+    };
+
+    Object.keys(thriftHandlers).forEach(function eachDef(def) {
+        tchannelAsThrift.register(tchannel, def, null,
+            thriftHandlers[def]);
+    });
+}
+
+function thriftCallback(callback) {
+    return function onCallback(err, res) {
+        if (err) {
+            callback(null, new ThriftError(err));
```

Comment:
this is not a valid thing to do. Please check `err.nameAsThrift` before creating a ThriftError.